src/hotspot/share/gc/z/zObjArrayAllocator.cpp

```diff
@@ -48,10 +48,11 @@ oop ZObjArrayAllocator::initialize(HeapWord* mem) const {
   // A max segment size of 64K was chosen because microbenchmarking
   // suggested that it offered a good trade-off between allocation
   // time and time-to-safepoint
-  const size_t segment_max = ZUtils::bytes_to_words(64 * K);
+  const size_t segment_max = 64 * K;
   const BasicType element_type = ArrayKlass::cast(_klass)->element_type();
-  const size_t header = arrayOopDesc::header_size(element_type);
-  const size_t payload_size = _word_size - header;
+  const size_t header = arrayOopDesc::base_offset_in_bytes(element_type);
+  size_t byte_size = _word_size * BytesPerWord;
+  const size_t payload_size = byte_size - header;
```

Comment:
Generational ZGC needs it to be able to write colored NULLs in object arrays, so a byte-centric approach will not work for us. Could you rewrite the code to take care of the "unaligned" 32 bits at the start of the array, and leave the word-centric segmented-clearing loop intact?

Reply:
Right. I restored the previous code and align-up the base offset. Object arrays in ZGC would not use the leading unaligned 32bits anyway, afaik.

Reply:
Note that even though the classes are named ObjArrayAllocator, this path is used for primitive arrays as well. I've wanted to change the name, but haven't gotten to it yet.

Reply:
Oh dear! Ok, I've added a section to properly clean the leading 32 bits if necessary, and verified by running some Z tests with -XX:-UseCompressedClassPointers. Thanks for pointing this out!